api/test/e2e/upstream_chash_hash_on_test.go

```diff
@@ -0,0 +1,379 @@
+	basepath := "http://127.0.0.1:9080"
+	var req *http.Request
+	var err error
+	var url string
+	var resp *http.Response
+	var respBody []byte
+	res := map[string]int{}
+	for i := 0; i <= 3; i++ {
+		url = basepath + "/server_port?var=2&var2=" + strconv.Itoa(i)
+		req, err = http.NewRequest("GET", url, nil)
+		req.Header.Add("custom_header", `custom-one`)
+		resp, err = http.DefaultClient.Do(req)
+		assert.Nil(t, err)
+		respBody, err = ioutil.ReadAll(resp.Body)
+		body := string(respBody)
+		if _, ok := res[body]; !ok {
+			res[body] = 1
+		} else {
+			res[body] += 1
+		}
+	}
+	assert.Equal(t, true, res["1980"] == 4 || res["1981"] == 4)
+	resp.Body.Close()
+}
+
+func TestUpstream_cHash_hash_on_cookie(t *testing.T) {
+	tests := []HttpTestCase{
+		{
+			caseDesc: "create cHash upstream with hash_on (cookie)",
+			Object:   ManagerApiExpect(t),
+			Method:   http.MethodPut,
+			Path:     "/apisix/admin/upstreams/1",
+			Body: `{
+					"nodes": [{
+						"host": "172.16.238.20",
+						"port": 1980,
+						"weight": 1
+					},
+					{
+						"host": "172.16.238.20",
+						"port": 1981,
+						"weight": 1
+					}],
+					"type": "chash",
+					"key": "custom-cookie",
+					"hash_on": "cookie"
+				}`,
+			Headers:      map[string]string{"Authorization": token},
+			ExpectStatus: http.StatusOK,
+		},
+		{
+			caseDesc: "create route using the upstream just created",
+			Object:   ManagerApiExpect(t),
+			Method:   http.MethodPut,
+			Path:     "/apisix/admin/routes/1",
+			Body: `{
+				"uri": "/server_port",
+				"upstream_id": "1"
+			}`,
+			Headers:      map[string]string{"Authorization": token},
+			ExpectStatus: http.StatusOK,
+			Sleep:        sleepTime,
+		},
+	}
+
+	for _, tc := range tests {
+		testCaseCheck(tc)
+	}
+
+	//hit routes
+	time.Sleep(time.Duration(100) * time.Millisecond)
+	basepath := "http://127.0.0.1:9080"
+	var req *http.Request
+	var err error
+	var url string
+	var resp *http.Response
+	var respBody []byte
+	res := map[string]int{}
+	for i := 0; i <= 3; i++ {
+		url = basepath + "/server_port?var=2&var2=" + strconv.Itoa(i)
+		req, err = http.NewRequest("GET", url, nil)
+		req.Header.Add("Cookie", `custom-cookie=cuscookie`)
+		resp, err = http.DefaultClient.Do(req)
+		assert.Nil(t, err)
+		respBody, err = ioutil.ReadAll(resp.Body)
+		body := string(respBody)
+		if _, ok := res[body]; !ok {
+			res[body] = 1
+		} else {
+			res[body] += 1
+		}
+	}
+	assert.Equal(t, true, res["1980"] == 4 || res["1981"] == 4)
+	resp.Body.Close()
+
+	//hit routes with miss cookie
+	res = map[string]int{}
+	for i := 0; i <= 3; i++ {
+		url = basepath + "/server_port?var=2&var2=" + strconv.Itoa(i)
+		req, err = http.NewRequest("GET", url, nil)
+		req.Header.Add("Cookie", `miss-custom-cookie=cuscookie`)
+		resp, err = http.DefaultClient.Do(req)
+		assert.Nil(t, err)
+		respBody, err = ioutil.ReadAll(resp.Body)
+		body := string(respBody)
+		if _, ok := res[body]; !ok {
+			res[body] = 1
+		} else {
+			res[body] += 1
+		}
+	}
```

Comment:
could use a same function with above cases.

Reply:
I will improve this in the future.